### PR TITLE
approval-distribution: a fix for out-of-view messages

### PR DESCRIPTION
**Pull request by ordian**
- [x] test on Versi
- [x] update the guide

### CODE REVIEW

File: node/network/approval-distribution/src/lib.rs [hunk 1]
```diff
@@ -995,43 +985,69 @@ impl State {
...
+							.collect();
+						// step 3.
+						// We assume if peer's knowledge is complete for block N,
+						// this is also true for its ancestors.
+						// This safeguard is needed primarily in case of long finality stalls
+						// so we don't waste time in a loop for every peer.
+						if missing.is_empty() {
+							return None
+						}
+						missing
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:1005]: nit: Adding a trace level log here might make sense

File: node/network/approval-distribution/src/lib.rs [hunk 2]
```diff
@@ -1048,8 +1064,20 @@ impl State {
...
+					let assignment_fingerprint = MessageFingerprint::Assignment(
+						block.clone(),
+						candidate_index,
+						validator_index.clone(),
+					);
+
 					match approval_state {
 						ApprovalState::Assigned(cert) => {
+							if !missing.contains(&assignment_fingerprint) {
+								continue
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:1076]: nit: Another tracing statement here would be well placed :)

**Comment by ordian** [comment]
Although it didn't eliminate the issue completely, we saw an improvement in the logs. And the remaining part is likely unrelated to the code here.

File: node/network/approval-distribution/src/lib.rs [hunk 3]
```diff
@@ -113,6 +113,8 @@ impl Knowledge {
 	}
 }
 
+type MissingKnowledge = HashSet<MessageFingerprint>;
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:116]: nit: A small descriptive comment would be appreciated, for this and the following type `PeerKnowledge`

File: node/network/approval-distribution/src/lib.rs [hunk 4]
```diff
@@ -1060,20 +1101,51 @@ impl State {
...
+									},
+									candidate_index.clone(),
+								));
+							} else {
+								tracing::trace!(
+									target: LOG_TARGET,
+									?block,
+									?validator_index,
+									?candidate_index,
+									"Skipping sending this known assignment",
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:1127]:
```suggestion
									"Skipping sending known assignment",
```

File: node/network/approval-distribution/src/lib.rs [hunk 5]
```diff
@@ -1060,20 +1101,51 @@ impl State {
...
+									candidate_index: candidate_index.clone(),
+									signature: signature.clone(),
+								});
+							} else {
+								tracing::trace!(
+									target: LOG_TARGET,
+									?block,
+									?validator_index,
+									?candidate_index,
+									"Skipping sending this known approval",
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:1146]:
```suggestion
									"Skipping sending known approval",
```

**Review by drahnr — APPROVED** [review]

File: node/network/approval-distribution/src/lib.rs [hunk 6]
```diff
@@ -113,8 +114,11 @@ impl Knowledge {
 	}
 }
 
+/// The difference of our knowledge and peer's knowledge
+/// that is used to send the missing information.
```
Comment (drahnr) [inline node/network/approval-distribution/src/lib.rs:118]: :heart:

**Comment by ordian** [comment]
bot merge

**Pull request merged**